### PR TITLE
More attribute refactoring

### PR DESCRIPTION
**Pull request by coolo**
Followup on #5158 - just to see regressions quicker

### CODE REVIEW

File: src/api/app/controllers/attribute_controller.rb [hunk 1]
```diff
@@ -2,14 +2,16 @@ class AttributeController < ApplicationController
...
-  validate_action attribute_definition: { method: :put, request: :attrib_type, response: :status }
-  validate_action attribute_definition: { method: :post, request: :attrib_type, response: :status }
+  validate_action show_namespace_definition: { method: :get, response: :attribute_namespace_meta }
+  validate_action delete_namespace_definition: { method: :delete, response: :status }
+  validate_action update_namespace_definition: { method: :put, request: :attribute_namespace_meta, response: :status }
+  validate_action update_namespace_definition: { method: :post, request: :attribute_namespace_meta, response: :status }
+  validate_action show_attribute_definition: { method: :get, response: :attrib_type }
+  validate_action delete_attribute_definition: { method: :delete, response: :status }
+  validate_action update_attribute_definition: { method: :put, request: :attrib_type, response: :status }
+  validate_action update_attribute_definition: { method: :post, request: :attrib_type, response: :status }
```
Comment (hennevogel) [inline src/api/app/controllers/attribute_controller.rb]: All of this sounds very much to me like two controllers, a `attribute_namespace` and a `attribute` controller.

And while you're at it, can you get rid of this stupid `_definition` prefix? :-)

File: src/api/app/controllers/attribute_controller.rb [hunk 2]
```diff
@@ -298,4 +244,51 @@ def find_attribute_container
     params[:namespace] = name_parts[0]
     params[:name] = name_parts[1]
   end
+
+  private
+
+  def ensure_namespace
```
Comment (hennevogel) [inline src/api/app/controllers/attribute_controller.rb]: This belongs to the routes. `namespace` is simply part of the route and if it is not there then this is a 404. No need to bother the controller with it.

File: src/api/app/controllers/attribute_controller.rb [hunk 3]
```diff
@@ -34,134 +36,76 @@ def index
     render xml: xml
   end
 
-  # /attribute/:namespace/_meta
-  def namespace_definition
-    if params[:namespace].nil?
-      raise MissingParameterError, "parameter 'namespace' is missing"
+  def show_namespace_definition
+    if (@an = AttribNamespace.where(name: ensure_namespace).select(:id, :name).first)
```
Comment (hennevogel) [inline src/api/app/controllers/attribute_controller.rb]: This sounds like a `require_namespace` `before_filter` you repeat this quite often below inside `ensure_namespace`. Please decouple this.

File: src/api/app/controllers/attribute_controller.rb [hunk 4]
```diff
@@ -34,134 +36,76 @@ def index
...
-    if request.get?
-      @an = AttribNamespace.where(name: namespace).select(:id, :name).first
-      unless @an
-        render_error message: "Unknown attribute namespace '#{namespace}'",
-          status: 404, errorcode: 'unknown_attribute_namespace'
-      end
-      return
-    end
+  def delete_namespace_definition
+    AttribNamespace.where(name: ensure_namespace).destroy_all
```
Comment (hennevogel) [inline src/api/app/controllers/attribute_controller.rb]: Please use `.find_by_name...` (inside `require_namespace`, see above) and `.destroy`

File: src/api/app/controllers/attribute_controller.rb [hunk 5]
```diff
@@ -34,134 +36,76 @@ def index
...
-        logger.debug '* create new attribute namespace'
-        AttribNamespace.create(name: namespace).update_from_xml(xml_element)
-      end
-
-      logger.debug '--- finished updating attribute namespace definitions ---'
-      render_ok
-    elsif request.delete?
-      AttribNamespace.where(name: namespace).destroy_all
-      render_ok
+    db = AttribNamespace.where(name: @namespace).first
```
Comment (hennevogel) [inline src/api/app/controllers/attribute_controller.rb]: While you're at it, please use speaking variable names.

**Comment by coolo** [comment]
I need #5181 to pass first - fighting rspec there

File: src/api/app/controllers/attribute_controller.rb (hunk 1 above)
Comment (coolo) [reply src/api/app/controllers/attribute_controller.rb]: It's actually 3 controllers - you notice when you try to remove the _definition suffix (prefix is ... anyway)

**Comment by coolo** [comment]
@hennevogel controllers are split - migrating the minitest to rspec is left to the interested reader :)

File: contrib/start_minitest [hunk 6]
```diff
@@ -5,7 +5,4 @@ echo "Preparing application..."
 bundle exec rake dev:bootstrap[old_test_suite] RAILS_ENV=test
 echo "Running api test suite..."
 export NO_MEMCACHED=1
-bundle exec rake test:api
-tmp_exit=$?
-rm -rf .bundle vendor/bundle
-exit $tmp_exit
+bundle exec ruby test/functional/attributes_test.rb # --name=test_create_namespace_old -v
```
Comment (coolo) [inline contrib/start_minitest]: oops - having an env variable to control this would be super cool. Having to patch local git sucks

**Comment by coolo** [comment]
@hennevogel please have another look - I would like to implement the actual feature I wanted to add. I just fell into the refactoring trap :)

File: src/api/app/mixins/has_attributes.rb [hunk 7]
```diff
@@ -40,20 +40,14 @@ def store_attribute(namespace, name, values, issues, binary = nil)
...
       # create the new attribute
       a = Attrib.new(attrib_type: attrib_type, binary: binary)
       a.project = self if is_a? Project
       a.package = self if is_a? Package
-      if a.attrib_type.value_count
-        a.attrib_type.value_count.times do |i|
-          a.values.build(position: i, value: values[i])
-        end
+      (a.attrib_type.value_count || 0).times do |i|
+        a.values.build(position: i, value: values[i])
```
Comment (bgeuken) [inline src/api/app/mixins/has_attributes.rb]: I am fine with the change, but this would change our API, right? I would at least document this in the release notes.

File: src/api/app/controllers/attribute_namespace_controller.rb [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+      AttribNamespace.create(name: @namespace).update_from_xml(xml_element)
+    end
+
+    render_ok
+  end
+
+  private
+
+  def require_namespace
+    @namespace = params[:namespace]
```
Comment (hennevogel) [inline src/api/app/controllers/attribute_namespace_controller.rb]: Why don't you do what the method name says and load the object here?

File: src/api/app/models/user.rb [hunk 9]
```diff
@@ -882,7 +872,7 @@ def hash_string(value)
                     'sha256crypt' => 5 }
     if deprecated_password_hash_type == 'md5'
       Digest::MD5.hexdigest(value + deprecated_password_salt)
-    elsif crypt2index.keys.include?(deprecated_password_hash_type)
+    elsif crypt2index.key?(deprecated_password_hash_type)
```
Comment (hennevogel) [inline src/api/app/models/user.rb]: What does this have to do with attributes?

File: src/api/app/mixins/has_attributes.rb (hunk 7 above)
Comment (coolo) [reply src/api/app/mixins/has_attributes.rb]: how so? by allowing value resorts you mean?

File: src/api/app/models/user.rb (hunk 9 above)
Comment (coolo) [reply src/api/app/models/user.rb]: atom beautifies the code on save

Comment (hennevogel) [reply src/api/app/models/user.rb]: Then atom should also make it an extra commit and describe why...

File: src/api/app/controllers/attribute_namespace_controller.rb (hunk 8 above)
Comment (coolo) [reply src/api/app/controllers/attribute_namespace_controller.rb]: because update and delete don't require the object but the name - you are free to continue the refactoring, but I spent enough time on it.

File: src/api/app/mixins/has_attributes.rb (hunk 7 above)
Comment (bgeuken) [reply src/api/app/mixins/has_attributes.rb]: I was referring to the whole [commit](995f93922874329940bca28dfcc7b6d9338a37ce), not that particular line. Just had to attach my comment somewhere.

But it looks like the API [changes](https://github.com/openSUSE/open-build-service/pull/5165/commits/d33b7c53015b4a1b4aad81fc2c3275af8fcbb206#diff-43a2207982cf719e0794f875b2b7c4e4R477). As said I personally think this is okay, since ithe change makes sense. But I would document this.

File: src/api/app/models/user.rb (hunk 9 above)
Comment (coolo) [reply src/api/app/models/user.rb]: Just as my cat owning neighbour said: I will talk to it, but I'm not sure it will change :)

**Comment by coolo** [comment]
For the API changes: I generally agree with you, but I wonder how to describe this. "If you use invalid URLs, be aware that we changed the reject reason" doesn't sound like something someone would want to read :)

**Comment by coolo** [comment]
reverted the robocop changes - will look at this cop in another PR

**Comment by coolo** [comment]
@hennevogel - squashed the fixes

File: src/api/test/functional/attributes_test.rb [hunk 10]
```diff
@@ -306,12 +306,13 @@ def test_attrib_delete_permissions
...
     data = "<attributes><attribute namespace='OBS' name='VeryImportantProject'/></attributes>"
     post '/source/home:tom/_attribute', params: data
+    assert_response :success
 
     login_tom
     delete '/source/home:tom/_attribute/OBS:VeryImportantProject'
     assert_response 403
     delete '/source/home:tom/_attribute/?namespace=OBS&name=VeryImportantProject'
-    assert_response 403
+    assert_response 404
```
Comment (adrianschroeter) [inline src/api/test/functional/attributes_test.rb]: Why is this a 404 now when it was not removed in the api call before?

Comment (coolo) [reply src/api/test/functional/attributes_test.rb]: because that route no longer exists.

Comment (coolo) [reply src/api/test/functional/attributes_test.rb]: I considered removing it - but I thought there is value in asserting you can't delete the attribute through the old route. But I'll do as you say.

Comment (coolo) [reply src/api/test/functional/attributes_test.rb]: See 97389774297e5426b2a20fa56d5ec39dece4c80c

**Comment by bgeuken** [comment]
> @coolo commented
> For the API changes: I generally agree with you, but I wonder how to describe this. "If you use invalid URLs, be aware that we changed the reject reason" doesn't sound like something someone would want to read :)

I would add to the [intentional changes](https://github.com/openSUSE/open-build-service/blob/master/ReleaseNotes-2.10#L55) section that:
* GET '/attribute/:attribute' route responded with a 400 when attribute did not exist. It no returns a 404 status.
* ...

File: src/api/app/controllers/attribute_namespace_controller.rb (hunk 8 above)
Comment (hennevogel) [reply src/api/app/controllers/attribute_namespace_controller.rb]: Okay

File: ReleaseNotes-2.10 [hunk 11]
```diff
@@ -82,6 +82,13 @@ Intentional changes:
       Rendered layouts/_google_analytics.html.erb (0.4ms)
     Completed 200 OK in 79ms (Views: 78.8ms | ActiveRecord: 0.0ms)
     ```
+ * In previous releases it was possible to delete attributes through
+   /source/<project>/_attribute/?namespace=OBS&name=VeryImportantProject (or similiar
+   for packages). You need to follow the documentation now and the proper
+   route is /source/<project>/_attribute/OBS:VeryImportantProject
+
+ * GET '/attribute/:attribute' route responded with a 400 when the attribute type
+   did not exist. It now returns a 404 status.
```
Comment (bgeuken) [inline ReleaseNotes-2.10]: @marcus-h osc does not rely on these endpoints, right? I just briefly checked osc command options...

**Comment by marcus-h** [comment]
On 2018-06-26 09:28:44 +0000, Björn Geuken wrote:
> bgeuken commented on this pull request.
> 
> 
> 
> > @@ -82,6 +82,13 @@ Intentional changes:
>        Rendered layouts/_google_analytics.html.erb (0.4ms)
>      Completed 200 OK in 79ms (Views: 78.8ms | ActiveRecord: 0.0ms)
>      ```
> + * In previous releases it was possible to delete attributes through
> +   /source/<project>/_attribute/?namespace=OBS&name=VeryImportantProject (or similiar
> +   for packages). You need to follow the documentation now and the proper
> +   route is /source/<project>/_attribute/OBS:VeryImportantProject
> +
> + * GET '/attribute/:attribute' route responded with a 400 when the attribute type
> +   did not exist. It now returns a 404 status.
> 
> @marcus-h osc does not rely on these endpoints, right? I just briefly checked osc command options...
> 
Yep, osc uses the
DELETE /source/<project>/_attribute/OBS:VeryImportantProject
route. Also, returning a 404 instead of a 400 sounds like a good
plan:)
From the osc POV everything's fine:)

**Comment by coolo** [comment]
still the flaky test - merging now after the 9th rebase :)

**Comment by hennevogel** [comment]
9 times red doesn't sound flaky to me...

**Comment by coolo** [comment]
Well, it's rather flaky green - but see https://github.com/openSUSE/open-build-service/pull/5188 for details
